### PR TITLE
Rebase SILScope generation on top of ASTScope.

### PR DESCRIPTION
**Pull request by adrian-prantl**
Rebase SILScope generation on top of ASTScope.

This patch replaces the stateful generation of SILScope information in SILGenFunction with data derived from the ASTScope hierarchy, which should be 100% in sync with the scopes needed for local variables. The goal is to eliminate the surprising effects that the stack of cleanup operations can have on the current state of SILBuilder leading to a fully deterministic (in the sense of: predictible by a human) association of SILDebugScopes with SILInstructions. The patch also eliminates the need to many workarounds. Thereare still some accomodations for several Sema transformation passes such as ResultBuilders, which don't correctly update the source locations when moving round nodes. If these were implemented as macros, this problem would disappear.

This necessary rewrite of the macro scope handling included in this patch also adds proper support nested macro expansions.

This fixes

rdar://88274783
and either fixes or at least partially addresses the following:

rdar://89252827
rdar://105186946
rdar://105757810
rdar://105997826
rdar://105102288

___

Cherry pick of #64551.

### CODE REVIEW

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/6611
@swift-ci test

File: lib/SILGen/SILGenEpilog.cpp [hunk 1]
```diff
@@ -319,6 +319,11 @@ emitEpilog(SILLocation TopLevel, bool UsesCustomEpilog) {
     // Return () if no return value was given.
     if (!returnValue)
       returnValue = emitEmptyTuple(CleanupLocation(TopLevel));
+    //{
+    //  RegularLocation loc(returnLoc);
+    //  loc.pointToEnd();
+    //  returnValue = emitEmptyTuple(loc);
+    //}
```
Comment (DougGregor) [inline lib/SILGen/SILGenEpilog.cpp:326]: Should this be uncommented or removed?

**Review by DougGregor — APPROVED** [review]

File: lib/SILGen/SILGenEpilog.cpp (hunk 1 above)
Comment (adrian-prantl) [reply lib/SILGen/SILGenEpilog.cpp:326]: Thanks for noticing!

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
Fix one more bug that was uncovered by the di-hole verification when building the source compatibility suite in debug mode.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test source compatibility

**Comment by adrian-prantl** [comment]
Source compatibility failures are unrelated to this PR.

**Pull request merged**